File: go/cmd/rekipedia/cmd/impact.go
package cmd

import (
	"fmt"
	"strings"

	"github.com/spf13/cobra"
	"github.com/unrealandychan/rekipedia/internal/storage"
)

var impactFlags struct {
	outputDir string
	depth     int
}

var impactCmd = &cobra.Command{
	Use:   "impact <file>",
	Short: "Show blast-radius for a changed file (BFS over reverse call graph)",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		targetFile := args[0]
		dbPath := storage.DefaultPath(impactFlags.outputDir)
		store, err := storage.Open(dbPath)
		if err != nil {
			return fmt.Errorf("open store: %w", err)
		}
		defer store.Close()

		runID, err := store.LatestRunID(impactFlags.outputDir)
		if err != nil || runID == "" {
			return fmt.Errorf("no scan found in %s — run reki scan first", impactFlags.outputDir)
		}

		symbols, err := store.ListSymbols(runID)
		if err != nil {
			return err
		}
		rels, err := store.ListRelationships(runID)
		if err != nil {
			return err
		}

		symFile := make(map[string]string)
		fileSyms := make(map[string][]string)
		for _, s := range symbols {
			symFile[s.Name] = s.File
			fileSyms[s.File] = append(fileSyms[s.File], s.Name)
		}

		reverse := make(map[string][]string)
		for _, r := range rels {
			if string(r.Kind) == "calls" {
				reverse[r.To] = append(reverse[r.To], r.From)
			}
		}

		seeds := fileSyms[targetFile]
		visited := make(map[string]bool)
		for _, s := range seeds {
			visited[s] = true
		}
		type qitem struct {
			sym   string
			depth int
		}
		queue := make([]qitem, 0)
		for _, s := range seeds {
			queue = append(queue, qitem{s, 0})
		}

		maxDepth := impactFlags.depth
		for len(queue) > 0 {
			cur := queue[0]
			queue = queue[1:]
			if cur.depth >= maxDepth {
				continue
			}
			for _, caller := range reverse[cur.sym] {
				if !visited[caller] {
					visited[caller] = true
					queue = append(queue, qitem{caller, cur.depth + 1})
				}
			}
		}

		seedSet := make(map[string]bool)
		for _, s := range seeds {
			seedSet[s] = true
		}

		affectedFiles := make(map[string]bool)
		var affectedSyms []string
		for sym := range visited {
			if seedSet[sym] {
				continue
			}
			affectedSyms = append(affectedSyms, sym)
			if f, ok := symFile[sym]; ok {
				affectedFiles[f] = true
			}
		}

		fmt.Printf("Impact: %s (depth=%d)\n", targetFile, maxDepth)
		fmt.Printf("Seed symbols: %d\n", len(seeds))
		fmt.Printf("Affected symbols: %d\n", len(affectedSyms))
		fmt.Printf("Affected files: %d\n", len(affectedFiles))

		var testFiles []string
		for f := range affectedFiles {
			if strings.Contains(f, "test") {
				testFiles = append(testFiles, f)
			}
		}
		if len(testFiles) > 0 {
			fmt.Printf("Related tests: %s\n", strings.Join(testFiles, ", "))
		}
		for f := range affectedFiles {
			fmt.Printf("  -> %s\n", f)
		}
		return nil
	},
}

func init() {
	impactCmd.Flags().StringVar(&impactFlags.outputDir, "output-dir", ".", "Directory with .rekipedia/")
	impactCmd.Flags().IntVar(&impactFlags.depth, "depth", 2, "BFS traversal depth")
}

File: src/rekipedia/cli/impact.py
"""reki impact — blast-radius analysis for a changed file."""
from __future__ import annotations
import click
from pathlib import Path
from rich.console import Console
from rich.tree import Tree

console = Console()

@click.command("impact")
@click.argument("target_file")
@click.option("--depth", default=2, show_default=True, help="BFS traversal depth")
@click.option("--output-dir", default=".", show_default=True, help="Directory with .rekipedia/")
def impact_cmd(target_file, depth, output_dir):
    """Show blast-radius for a changed file."""
    from rekipedia.storage.sqlite_store import SqliteStore
    from rekipedia.analysis.impact import compute_impact

    db_path = Path(output_dir) / ".rekipedia" / "store.db"
    if not db_path.exists():
        alt = Path(output_dir) / ".rekipedia" / "rekipedia.db"
        if alt.exists():
            db_path = alt
    if not db_path.exists():
        console.print(f"[red]No rekipedia DB at {Path(output_dir) / '.rekipedia' / 'store.db'}. Run reki scan first.[/red]")
        raise click.Abort()

    store = SqliteStore(db_path)
    run_id = store.latest_run_id()
    if not run_id:
        console.print("[red]No scan runs found.[/red]")
        raise click.Abort()

    symbols = store.get_all_symbols(run_id)
    relationships = store.get_all_relationships(run_id)

    result = compute_impact(target_file, relationships, symbols, depth=depth)

    tree = Tree(f"[bold cyan]Impact: {target_file}[/bold cyan] (depth={depth})")
    affected = tree.add(f"[yellow]Affected files ({len(result['affected_files'])})[/yellow]")
    for f in result["affected_files"]:
        affected.add(f)
    tests = tree.add(f"[green]Related tests ({len(result['related_tests'])})[/green]")
    for t in result["related_tests"]:
        tests.add(t)
    syms = tree.add(f"[blue]Affected symbols ({result['total_affected']})[/blue]")
    for s in result["affected_symbols"][:20]:
        syms.add(s)
    if result["total_affected"] > 20:
        syms.add(f"... and {result['total_affected'] - 20} more")
    console.print(tree)
